src/main/java/de/tum/in/www1/artemis/service/ExamService.java

```diff
@@ -182,10 +187,148 @@ public void delete(Long examId) {
+     * @return optional of relevant result
+     */
+    private Optional<Result> getRelevantResult(StudentParticipation studentParticipation) {
+        // no participant -> no relevant result
+        if (studentParticipation.getParticipant() == null) {
+            return Optional.empty();
+        }
+
+        return studentParticipation.getResults().stream().filter(Result::isRated).filter(result -> result.getCompletionDate() != null).filter(result -> result.getScore() != null)
+                .sorted((r1, r2) -> r2.getCompletionDate().compareTo(r1.getCompletionDate())).findFirst();
+
+    }
+
+    /**
+     * Rounds a double to a specified number of decimal places
+     *
+     * @param value  the number to round
+     * @param places the number of decimal places to round to
+     * @return rounded number
+     */
+    public static double round(double value, int places) {
+        if (places < 0) {
+            throw new IllegalArgumentException();
+        }
+
+        BigDecimal bd = BigDecimal.valueOf(value);
+        bd = bd.setScale(places, RoundingMode.HALF_UP);
+        return bd.doubleValue();
+    }
+
+    /**
+     * Puts students, result and exerciseGroups together for ExamScoresDTO
+     *
+     * @param examId the id of the exam
+     * @return return ExamScoresDTO with students, scores and exerciseGroups for exam
+     */
+    public ExamScoresDTO getExamScore(Long examId) {
+        Exam exam = examRepository.findForScoreCalculationById(examId).orElseThrow(() -> new EntityNotFoundException("Exam with id: \"" + examId + "\" does not exist"));
+
+        // Adding exam information to DTO
+        ExamScoresDTO scores = new ExamScoresDTO(exam.getId(), exam.getTitle(), exam.getMaxPoints());
+
+        // Adding exercise group information to DTO
+        for (ExerciseGroup exerciseGroup : exam.getExerciseGroups()) {
+            // Alert: This only works if all exercises in an exercise groups have the same number of maximum points
+            Double maximumNumberOfPoints = null;
+            if (!exerciseGroup.getExercises().isEmpty()) {
+                maximumNumberOfPoints = exerciseGroup.getExercises().iterator().next().getMaxScore();
+            }
+
+            List<String> containedExercises = new ArrayList<>();
+
+            for (Exercise exercise : exerciseGroup.getExercises()) {
+                containedExercises.add(exercise.getTitle().trim());
+            }
+
+            scores.exerciseGroups.add(new ExamScoresDTO.ExerciseGroup(exerciseGroup.getId(), exerciseGroup.getTitle(), maximumNumberOfPoints, containedExercises));
+        }
+
+        // Adding registered student information to DTO
+        for (User user : exam.getRegisteredUsers()) {
+            scores.studentResults.add(new ExamScoresDTO.StudentResult(user.getId(), user.getName(), user.getEmail(), user.getLogin(),
+                    user.getRegistrationNumber() != null ? user.getRegistrationNumber().trim() : ""));
```

Comment:
Why do we return "" instead of null if no registrationNumber is present?

Reply:
yes, we should return null, trim() is also not necessary